core/src/main/java/amino/run/oms/migrationDecision/MetricWatcher.java

```diff
@@ -0,0 +1,287 @@
+package amino.run.oms.migrationDecision;
+
+import amino.run.common.MicroServiceID;
+import amino.run.common.MicroServiceNotFoundException;
+import amino.run.common.MicroServiceReplicaNotFoundException;
+import amino.run.common.ReplicaID;
+import amino.run.kernel.common.KernelOID;
+import amino.run.kernel.common.KernelServerNotFoundException;
+import amino.run.kernel.common.metric.NodeMetric;
+import amino.run.kernel.common.metric.RPCMetric;
+import amino.run.oms.KernelServerManager;
+import amino.run.oms.MicroServiceManager;
+import amino.run.policy.DefaultPolicy;
+import amino.run.policy.MigrationNotification;
+import amino.run.policy.util.ResettableTimer;
+import amino.run.runtime.EventHandler;
+import java.net.InetSocketAddress;
+import java.rmi.RemoteException;
+import java.util.*;
+import java.util.logging.Logger;
+
+/**
+ * Class watch over metric collected in {@link MicroServiceManager} and {@link KernelServerManager}.
+ * For evaluating best candidate for migration it maintains a heap based on time consumption in
+ * Micro service replica.
+ *
+ * @author AmitRoushan
+ */
+public class MetricWatcher {
+    private Logger logger = Logger.getLogger(MetricWatcher.class.getName());
+    private MicroServiceManager microServiceManager;
+    private KernelServerManager serverManager;
+    private MigrationPredictor predictor;
+    private transient volatile ResettableTimer timer;
+    private TreeSet<AbstractMap.SimpleEntry<ReplicaID, Long>> replicaHeap;
+    private final int MAX_TOP_CANDIDATE = 3;
+    /** Time interval of metric watch timer */
+    public final int METRIC_WATCH_INTERVAL = 30000; // in milli seconds
+
+    public MetricWatcher(
+            KernelServerManager serverManager, MicroServiceManager microServiceManager) {
+        this.microServiceManager = microServiceManager;
+        this.serverManager = serverManager;
+        this.predictor = new MigrationPredictor(serverManager, microServiceManager);
+        replicaHeap =
+                new TreeSet<AbstractMap.SimpleEntry<ReplicaID, Long>>(
+                        new Comparator<AbstractMap.SimpleEntry<ReplicaID, Long>>() {
+                            @Override
+                            public int compare(
+                                    AbstractMap.SimpleEntry<ReplicaID, Long> replicaStat1,
+                                    AbstractMap.SimpleEntry<ReplicaID, Long> replicaStat2) {
+                                if (replicaStat1.getValue() < replicaStat2.getValue()) return 1;
+                                else if (replicaStat1.getValue() > replicaStat2.getValue())
+                                    return -1;
+                                return 0;
+                            }
+                        });
+    }
+
+    /** Starts metric watcher */
+    public void start() {
+        if (timer == null) {
```

Comment:
Instead if timer != null, then simply return can be more clear ?

Reply:
Addressed in #793 PR